src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -151,10 +151,11 @@ internal void RuntimeNotify(byte[] eventName, Array state)
         {
             if (eventName.Length > MaxEventName) throw new ArgumentException();
             if (!CheckItemForNotification(state)) throw new ArgumentException();
-            SendNotification(CurrentScriptHash, Encoding.UTF8.GetString(eventName), state);
+            if (state.Any(p => !p.IsNull && !(p is PrimitiveType))) throw new ArgumentException();
```

Comment:
we can't send an array?

Reply:
I don't think it's worth restricting this interface that much, we should just make a deep copy of notification's stack item and make a copy again if anyone is to retreive it via `GetNotifications`, that's the behavior I would expect.

Reply:
> I don't think it's worth restricting this interface that much, we should just make a deep copy of notification's stack item and make a copy again if anyone is to retreive it via GetNotifications, that's the behavior I would expect.

Agree

Reply:
Okay, I will modify it.

Reply:
> I don't think it's worth restricting this interface that much, we should just make a deep copy of notification's stack item and make a copy again if anyone is to retreive it via GetNotifications, that's the behavior I would expect.

The first copy is not necessary, right?

Reply:
> The first copy is not necessary, right?

I think it is necessary, when a contract emits something it must be immutable, like a log line, when you're logging something you're not supposed to change this line afterwards, so there shouldn't be any possibility of doing that. And notifications are a part of contract's API, so any receivers of these notifications (smart contract backends mostly) should have some guarantees of them being intact, delivered exactly as the contract emitted them.

Reply:
But the only chance you can modify it is `GetNotifications`.

Reply:
I don't think so, there also a possibility for things like this:
```
[TestMethod]
public void Runtime_GetNotifications_Test()
{
        var snapshot = Blockchain.Singleton.GetSnapshot();
        using (var engine = new ApplicationEngine(TriggerType.Application, null, snapshot, 0, true))
        using (var script = new ScriptBuilder())
        {
            // Notification                                                                                                                                                                                                                                                                                                                                                                  
            script.EmitPush("eventPayload");
            script.EmitPush(1);
            script.Emit(OpCode.PACK);
            script.Emit(OpCode.DUP); // save array                                                                                                                                                                                                                                                                                                                                           
            script.EmitPush("eventName");
            script.EmitSysCall(ApplicationEngine.System_Runtime_Notify);
            script.EmitPush(0);
            script.EmitPush("modifiedEvent");
            script.Emit(OpCode.SETITEM);
            engine.LoadScript(script.ToArray());
            Assert.AreEqual(VMState.HALT, engine.Execute());
            Assert.AreEqual(1, engine.Notifications.Count);
            var array = (VM.Types.Array)engine.Notifications[0].State;
            Assert.AreEqual("eventPayload", array[0].GetString());
        }
}
```
And at the moment it works this way:
```
   Assert.AreEqual failed. Expected:<eventPayload>. Actual:<modifiedEvent>. 
```
Which I think is not correct and even dangerous (this array may be passed between contracts, for example).